app/code/Magento/Catalog/Model/Product/Gallery/UpdateHandler.php

```diff
@@ -21,16 +21,27 @@ protected function processDeletedImages($product, array &$images)
         $recordsToDelete = [];
         $picturesInOtherStores = [];
 
-        foreach ($this->resourceModel->getProductImages($product, $this->extractStoreIds($product)) as $image) {
-            $picturesInOtherStores[$image['filepath']] = true;
+        $galleryImages = $this->resourceModel->getProductGallery(
+            $product,
+            $this->getAttribute()->getAttributeId(),
+            $this->extractStoreIds($product)
+        );
+        foreach ($galleryImages as $image) {
+            // Same file can share among gallery on store level
+            $picturesInOtherStores[$image['file']] = true;
         }
 
         foreach ($images as &$image) {
             if (!empty($image['removed'])) {
-                if (!empty($image['value_id']) && !isset($picturesInOtherStores[$image['file']])) {
+                if (!empty($image['value_id'])) {
+                    /**
+                     * Records work on store level and should be deleted no matter the same file
+                     * used across stores or not
+                     */
                     $recordsToDelete[] = $image['value_id'];
+
                     // only delete physical files if they are not used by any other products
-                    if (!($this->resourceModel->countImageUses($image['file']) > 1)) {
+                    if (!isset($picturesInOtherStores[$image['file']])) {
```

Comment:
As now `catalog_product_entity_media_gallery` work on store level, `countImageUses` cannot delete the image when it is used by two gallery entries on the same store and getting deleted at the same time.